crates/ironrdp-connector/src/credssp_sequence.rs

```diff
@@ -0,0 +1,214 @@
+};
+
+#[derive(Clone, Copy, Debug)]
+struct CredsspTsRequestHint;
+
+const CREDSSP_TS_REQUEST_HINT: CredsspTsRequestHint = CredsspTsRequestHint;
+
+impl PduHint for CredsspTsRequestHint {
+    fn find_size(&self, bytes: &[u8]) -> ironrdp_pdu::PduResult<Option<usize>> {
+        match sspi::credssp::TsRequest::read_length(bytes) {
+            Ok(length) => Ok(Some(length)),
+            Err(e) if e.kind() == std::io::ErrorKind::UnexpectedEof => Ok(None),
+            Err(e) => Err(ironrdp_pdu::custom_err!("CredsspTsRequestHint", e)),
+        }
+    }
+}
+#[derive(Clone, Copy, Debug)]
+struct CredsspEarlyUserAuthResultHint;
+
+const CREDSSP_EARLY_USER_AUTH_RESULT_HINT: CredsspEarlyUserAuthResultHint = CredsspEarlyUserAuthResultHint;
+
+impl PduHint for CredsspEarlyUserAuthResultHint {
+    fn find_size(&self, _: &[u8]) -> ironrdp_pdu::PduResult<Option<usize>> {
+        Ok(Some(sspi::credssp::EARLY_USER_AUTH_RESULT_PDU_SIZE))
+    }
+}
+
+pub type CredSspProcessGenerator<'a> = Generator<'a, NetworkRequest, sspi::Result<Vec<u8>>, sspi::Result<ClientState>>;
+
+#[derive(Debug)]
+pub struct CredSspSequence {
+    client: CredSspClient,
+    next_request: Option<credssp::TsRequest>,
+    state: CredSSPState,
+    selected_protocol: nego::SecurityProtocol,
+}
+
+#[derive(Debug, PartialEq)]
+pub(crate) enum CredSSPState {
+    CredsspInitial,
+    CredsspReplyNeeded,
+    CredsspEarlyUserAuthResult,
+    Finished,
+}
+
+impl CredSspSequence {
+    pub fn next_pdu_hint(&self) -> Option<&dyn PduHint> {
+        match self.state {
+            CredSSPState::CredsspInitial => None,
+            CredSSPState::CredsspReplyNeeded => Some(&CREDSSP_TS_REQUEST_HINT),
+            CredSSPState::CredsspEarlyUserAuthResult => Some(&CREDSSP_EARLY_USER_AUTH_RESULT_HINT),
+            CredSSPState::Finished => None,
+        }
+    }
+
+    pub fn new(
+        connector: &ClientConnector,
+        server_name: ServerName,
+        server_public_key: Vec<u8>,
+        kerberos_config: Option<KerberosConfig>,
+    ) -> ConnectorResult<Self> {
+        let config = &connector.config;
+        if let crate::Credentials::SmartCard { .. } = config.credentials {
+            return Err(general_err!(
+                "CredSSP with smart card credentials is not currently supported"
+            ));
+        }
+
+        let credentials = sspi::AuthIdentity {
+            username: config.credentials.username().into(),
+            password: config.credentials.secret().to_owned().into(),
+            domain: config.domain.clone(),
+        };
+
+        let server_name = server_name.into_inner();
+
+        let service_principal_name = format!("TERMSRV/{}", &server_name);
+
+        let credssp_config: Box<dyn ProtocolConfig>;
+        if let Some(ref krb_config) = kerberos_config {
+            credssp_config = Box::new(Into::<SSPIKerberosConfig>::into(krb_config.clone()));
+        } else {
+            credssp_config = Box::<sspi::ntlm::NtlmConfig>::default();
+        }
```

Comment:
We should use the "negotiate" security provider: https://github.com/Devolutions/sspi-rs/blob/7779af59d32af845f25778897dc5ff8e5febde2f/src/negotiate.rs#L37-L42
https://learn.microsoft.com/en-us/windows/win32/secauthn/microsoft-negotiate

This provider will automatically use Kerberos and fallback to NTLM when necessary.

Reply:
yes, that is indeed what I use. it takes a protocol config as parameter, which if it is Kerbero config, it will user Kerberos, then if it fails, it will fall back to NTLM, if Kerberos config not persented at all, it starts with NTLM. 
``` Rust
 let credssp_config: Box<dyn ProtocolConfig>;
        if let Some(ref krb_config) = kerberos_config {
            credssp_config = Box::new(Into::<sspi::KerberosConfig>::into(krb_config.clone()));
        } else {
            credssp_config = Box::<sspi::ntlm::NtlmConfig>::default();
        }
        
        info!("using config : {:?}", &credssp_config);

        let client = credssp::CredSspClient::new(
            server_public_key,
            credentials.into(),
            credssp::CredSspMode::WithCredentials,
            credssp::ClientMode::Negotiate(sspi::NegotiateConfig { //<-- here
                protocol_config: credssp_config,
                package_list: None,
                hostname: server_name,
            }),
            service_principal_name,
        )
```

Reply:
Oh, okay! Sounds good